### PR TITLE
fix quadratic time in `is_cyclic` and `topological_sort` (fix #263)

### LINKED ISSUE #263: Possible issue with topological_sort_by_dfs
Looks like the function ```topological_sort_by_dfs``` can have a run-time $O(n\cdot n)$ even for sparse graphs. The simplest example to get this behaviour is

```
function gen_graph_topsort(k)
    g = SimpleDiGraph(k+1)
    for i = 1:k
        add_edge!(g, 1, i+1)
    end
    return g
end

g = gen_graph_topsort(10^5)
@time topological_sort_by_dfs(g)
```

For graphs with about 10^5 vertices this already takes a few seconds. For larger graphs, this quickly becomes infeasible (due to the quadratic run-time). Expected behaviour would be that the implementation has worst-case O(n+m) run-time and takes fractions of a second for graphs of this size.

The problem appears to be that this loop https://github.com/JuliaGraphs/Graphs.jl/blob/e5405d105e034ccb9396d30f86c26641ce94c504/src/traversals/dfs.jl#L97 goes through the neighbors of u over and over everytime starting from the beginning of the list.

I.e., it will find ```n``` with ```vcolor[n] == 0``` and break, then (with some further steps) ```vcolor[n]``` will be set to ```2```. Afterwards, again the search over the neighbors of u starts at the first neighbor and goes through them until it finds one with ```vcolor == 0``` and breaks etc.

Thus, for a vertex with $k$ neighbor it can take $O(k^2)$ steps, leading to potentially O(n^2), resp. O(m^2), run-time for sparse graphs.

### PR DESCRIPTION
**Pull request by etiennedeg**
Fixes #263

### CODE REVIEW

File: src/traversals/dfs.jl [hunk 1]
```diff
@@ -33,26 +33,25 @@ end
 # see https://github.com/mauro3/SimpleTraits.jl/issues/47#issuecomment-327880153 for syntax
 @traitfn function is_cyclic(g::AG::IsDirected) where {T,AG<:AbstractGraph{T}}
     vcolor = zeros(UInt8, nv(g))
+    S = Vector{T}()
```
Comment (gdalle) [inline src/traversals/dfs.jl:36]: Comment what `S` stands for?

File: src/traversals/dfs.jl [hunk 2]
```diff
@@ -87,32 +86,31 @@ function topological_sort_by_dfs end
 @traitfn function topological_sort_by_dfs(g::AG::IsDirected) where {T,AG<:AbstractGraph{T}}
     vcolor = zeros(UInt8, nv(g))
     verts = Vector{T}()
+    S = Vector{T}()
```
Comment (gdalle) [inline src/traversals/dfs.jl:89]: Comment what `S` stands for?

File: src/traversals/dfs.jl [hunk 3]
```diff
@@ -33,26 +33,25 @@ end
 # see https://github.com/mauro3/SimpleTraits.jl/issues/47#issuecomment-327880153 for syntax
 @traitfn function is_cyclic(g::AG::IsDirected) where {T,AG<:AbstractGraph{T}}
     vcolor = zeros(UInt8, nv(g))
```
Comment (gdalle) [inline src/traversals/dfs.jl:35]: Comment what the color values stand for?

File: src/traversals/dfs.jl [hunk 4]
```diff
@@ -87,32 +86,31 @@ function topological_sort_by_dfs end
 @traitfn function topological_sort_by_dfs(g::AG::IsDirected) where {T,AG<:AbstractGraph{T}}
     vcolor = zeros(UInt8, nv(g))
```
Comment (gdalle) [inline src/traversals/dfs.jl:87]: Comment what the color values stand for?

File: src/traversals/dfs.jl [hunk 5]
```diff
@@ -87,32 +86,31 @@ function topological_sort_by_dfs end
...
-            end
-            if w != 0
-                vcolor[w] = 1
-                push!(S, w)
             else
-                vcolor[u] = 2
-                push!(verts, u)
+                vcolor[u] = 3
                 pop!(S)
+                pushfirst!(verts, u)
```
Comment (gdalle) [inline src/traversals/dfs.jl:109]: How are we sure that each vertex lands in `verts` exactly once?

**Review by gdalle — CHANGES_REQUESTED** [review]

File: src/traversals/dfs.jl (hunk 5 above)
Comment (etiennedeg) [reply src/traversals/dfs.jl:115]: Each vertex is visited at least once (outer loop). Each vertex is visited at most once (added to vertex_heap only if color is 0).
Then vertex is visited once to explore its unvisited subtree, and color goes to 2. Next time it is visited, its color goes to 3, and vertex is removed from queue and added to verts, so not visited anymore.

File: src/traversals/dfs.jl [hunk 6]
```diff
@@ -32,28 +32,30 @@ function is_cyclic end
 end
 # see https://github.com/mauro3/SimpleTraits.jl/issues/47#issuecomment-327880153 for syntax
 @traitfn function is_cyclic(g::AG::IsDirected) where {T,AG<:AbstractGraph{T}}
+    # 0 if not visited, 1 if visited, 2 if in the current dfs path, 3 if fully explored 
     vcolor = zeros(UInt8, nv(g))
+    vertex_queue = Vector{T}()
```
Comment (gdalle) [inline src/traversals/dfs.jl:37]: Why a Vector instead of another data structure?

Comment (etiennedeg) [reply src/traversals/dfs.jl:37]: I kept the last one, why do you want something else ? We only need to push and pop, there are no weight ordering

Comment (simonschoelly) [reply src/traversals/dfs.jl:37]: There should be some queue data structures in [DataStructures.jl](https://github.com/JuliaCollections/DataStructures.jl) - but I am not sure if and how much more efficient they would be. But if I understand correctly, you always add or remove at the end, so it should be called `stack` not `queue`.

Comment (gdalle) [reply src/traversals/dfs.jl:37]: I think we should merge this as is (with a renaming) and keep track of the potential for a better stack in a new issue?

**Review by gdalle — APPROVED** [review]
Seems fine to me

**Pull request merged**